prboom2/src/g_game.c

```diff
@@ -3179,13 +3182,74 @@ void G_BeginRecording (void)
 {
   int i;
   byte *demostart, *demo_p;
+  int num_extensions = 0;
   demostart = demo_p = malloc(1000);
   longtics = 0;
 
-  if (umapinfo_loaded)
+  // ano - jun2019 - add the extension format if needed
+  if (gamemapinfo)
+  {
+    num_extensions++;
+  }
+
+  if (num_extensions > 0)
   {
-	  *demo_p++ = 255;
+    // demover
+    *demo_p++ = 0xFF;
+    // signature
+    *demo_p++ = prdemosig[0]; // 'P'
+    *demo_p++ = prdemosig[1]; // 'R'
+    *demo_p++ = prdemosig[2]; // '+'
+    *demo_p++ = prdemosig[3]; // 'U'
+    *demo_p++ = prdemosig[4]; // 'M'
+    *demo_p++ = '\0';
+    // extension version
+    *demo_p++ = 1;
+    //
+    *demo_p++ =  num_extensions & 0xff;
+    *demo_p++ = (num_extensions >> 8) & 0xff;
+
+    if (gamemapinfo)
+    {
+      int mapname_len = strnlen(gamemapinfo->mapname, 9);
+
+      // ano - note that the format has each length by each string
+      // as opposed to a table of lengths
+      *demo_p++ = 0x08;
+      *demo_p++ = 'U';
+      *demo_p++ = 'M';
+      *demo_p++ = 'A';
+      *demo_p++ = 'P';
+      *demo_p++ = 'I';
+      *demo_p++ = 'N';
+      *demo_p++ = 'F';
+      *demo_p++ = 'O';
+
+      // ano - to properly extend this to support other extension strings
+      // we wouldn't just plop this here, but right now we only support the 1
+      // in the future, we should assume that chunks in the header should
+      // follow the order of their appearance in the extensions table.
+      if (mapname_len > 8)
+      {
+        I_Error("Unable to save map lump name %s, too large.", gamemapinfo->mapname);
+      }
+
+      for (i = 0; i < mapname_len; i++)
+      {
+        // FIXME - the toupper is a hacky workaround for the case insensitivity
+        // in the current UMAPINFO reader. lump names should probably not be
+        // lowercase ever (?)
+        *demo_p++ = toupper(gamemapinfo->mapname[i]);
```

Comment:
Why do we store the map lump name here? It's also stored in the regular demo header and at this point it makes no difference if the level is loaded by means of UMAPINFO or not - it's just the level progression that matters. To make sure that this gets right, we need to enforce the UMAPINFO lump getting loaded - which is the entire purpose of this new header.

Reply:
OK I've consulted with anotak on the matter. Turns out some demo formats only support map in the form of numbers. This change means we cah have demos without the gameplay you'd have for those formats, but with UMAPINFO support.

Reply:
Ye, basically it's a bit of forward-compatibility, in case UMAPINFO grows to support non-standard map lump names (e.g. "E1A1"). Without this, it'd be impossible to, say, record a -complevel 2 demo in a wad that uses a nonstandard map lump names but is otherwise intended for limit-removing play... and by "impossible" I mean "would fail wildly". :P

Besides, even if we assume the longest case (9 bytes per map lump name) for someone recording a full-megawad playthrough demo of 32 maps, this would only be 288 bytes, which is already less character than this message and is orders of magnitude less data than what you've downloaded by viewing this comment on GitHub. :P

tl;dr, it's necessary to make things future-proof and the extra space it takes is friggin' tiny in 2020.